polynote-spark-runtime/src/main/scala/polynote/runtime/spark/reprs/SparkReprsOf.scala

```diff
@@ -3,74 +3,181 @@ package polynote.runtime.spark.reprs
+      arr
+    }
+  }
 
-      StructType(fieldTypes.toList) -> {
-        (out, row) =>
-          fieldEncoders.foreach(_.apply(out, row))
+  private[polynote] class VariableSizeDataFrameDecoder(structType: StructType, encode: (DataOutput, InternalRow) => Unit) extends (InternalRow => Array[Byte]) with Serializable {
+    override def apply(row: InternalRow): Array[Byte] = {
+      val out = new ByteArrayOutputStream()
+      try {
+        encode(new DataOutputStream(out), row)
+        out.toByteArray
+      } finally {
+        out.close()
       }
     }
+  }
 
-    instance {
-      df =>
-        val (structType, encode) = dataTypeAndEncoder(df.schema)
-        val rowToBytes: InternalRow => ByteBuffer = if (structType.size >= 0) {
-          row: InternalRow =>
-            // TODO: share/reuse this buffer? Any reason to be threadsafe?
-            val buf = ByteBuffer.allocate(structType.size)
-            encode(new DataEncoder.BufferOutput(buf), row)
-            buf
-        } else {
-          row: InternalRow =>
-            val out = new ByteArrayOutputStream()
-            try {
-              encode(new DataOutputStream(out), row)
-              ByteBuffer.wrap(out.toByteArray)
-            } finally {
-              out.close()
+
+  private[polynote] class DataFrameHandle(
+    val handle: Int,
+    dataFrame: DataFrame
+  ) extends StreamingDataRepr.Handle with Serializable {
+
+    private val originalStorage = dataFrame.storageLevel
+    private val (structType, encode) = dataTypeAndEncoder(dataFrame.schema)
+
+
+    val dataType: DataType = structType
+    val knownSize: Option[Int] = None
+
+    // TODO: It might be nice to iterate instead of collect, but maybe in a better way than toLocalIterator...
+    private lazy val collectedData = {
+      val rowToBytes: InternalRow => Array[Byte] = if (structType.size >= 0) {
+        new FixedSizeDataFrameDecoder(structType, encode)
+      } else {
+        new VariableSizeDataFrameDecoder(structType, encode)
+      }
+
+      dataFrame.limit(1000000).queryExecution.toRdd.map(rowToBytes).collect()
+    }
+
+    def iterator: Iterator[ByteBuffer] = collectedData.iterator.map(ByteBuffer.wrap)
+
+
+    def modify(ops: List[TableOp]): Either[Throwable, Int => StreamingDataRepr.Handle] = {
+      import org.apache.spark.sql.functions, functions.{when, col, sum, lit, struct, count, approx_count_distinct, avg}
+      import org.apache.spark.sql.Column
+      import org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute
+      import org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentile
+      import org.apache.spark.sql.catalyst.expressions.Literal
+
+
+      @inline def tryEither[T](thunk: => T): Either[Throwable, T] = try Right(thunk) catch {
+        case err: Throwable => Left(err)
+      }
+
+      def toAggregate(strings: (String, String)): (List[Column], Option[DataFrame => DataFrame]) = strings match {
```

Comment:
Note, all of these aggregates ought to also be implemented for i.e. list-of-case-classes. TODO.